File: Emolga/src/de/tectoast/emolga/commands/showdown/SearchReplaysCommand.java
package de.tectoast.emolga.commands.showdown;

import de.tectoast.emolga.commands.Command;
import de.tectoast.emolga.commands.CommandCategory;
import de.tectoast.emolga.commands.GuildCommandEvent;
import net.dv8tion.jda.api.entities.TextChannel;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.IOException;
import java.net.URL;

public class SearchReplaysCommand extends Command {
    public SearchReplaysCommand() {
        super("searchreplays", "`!searchreplays <User1> [User2]` Sucht nach Replays der angegebenen Showdownbenutzernamen", CommandCategory.Showdown);
    }

    @Override
    public void process(GuildCommandEvent e) throws IOException {
        TextChannel tco = e.getChannel();
        String msg = e.getMessage().getContentDisplay().substring(15);
        String url = "https://replay.pokemonshowdown.com/search.json?user=";
        String[] split = msg.split(" ");
        if(split.length > 1) url += split[0].toLowerCase() + "&user2=" + split[1].toLowerCase();
        else url += split[0];
        System.out.println(url);
        JSONArray array = new JSONArray(new JSONTokener(new URL(url).openStream()));
        System.out.println(array.toString(4));
        StringBuilder str = new StringBuilder();
        if(array.length() == 0) {
            tco.sendMessage("Es wurde kein Kampf zwischen " + split[0] + " und " + split[1] + " hochgeladen!").queue();
            return;
        }
        for (int i = 0; i < array.length(); i++) {
            JSONObject o = array.getJSONObject(i);
            str.append(o.getString("p1")).append(" vs ").append(o.getString("p2")).append(": https://replay.pokemonshowdown.com/").append(o.getString("id")).append("\n");
        }
        System.out.println(str.toString());

        e.getChannel().sendMessage(str.toString()).queue();
    }
}

File: src/main/kotlin/de/tectoast/emolga/features/showdown/SearchReplaysCommand.kt
package de.tectoast.emolga.features.showdown

import de.tectoast.emolga.features.Arguments
import de.tectoast.emolga.features.CommandFeature
import de.tectoast.emolga.features.CommandSpec
import de.tectoast.emolga.features.InteractionData
import de.tectoast.emolga.utils.httpClient
import de.tectoast.emolga.utils.notNullAppend
import de.tectoast.emolga.utils.otherJSON
import de.tectoast.emolga.utils.toUsername
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.coroutines.delay
import kotlinx.serialization.Serializable
import mu.KotlinLogging

object SearchReplaysCommand : CommandFeature<SearchReplaysCommand.Args>(
    ::Args,
    CommandSpec(
        "searchreplays",
        "Sucht nach Replays der angegebenen Showdownbenutzernamen",
    )
) {
    class Args : Arguments() {
        var user1 by string("User 1", "Der Showdown-Username von jemandem")
        var user2 by string("User 2", "Der Showdown-Username eines potenziellen zweiten Users").nullable()
    }

    context(iData: InteractionData)
    override suspend fun exec(e: Args) {
        val u1 = e.user1
        val u2 = e.user2
        iData.deferReply()
        val body = getBody(u1, u2)
        logger.info(body)
        val jsonstring = if (body.length < 30) {
            iData.reply("Verbindung mit dem Showdown-Server fehlgeschlagen, ich versuche es in 10 Sekunden erneut...")
            delay(10000)
            getBody(u1, u2)
        } else body
        try {
            iData.reply(
                otherJSON.decodeFromString<List<Replay>>(jsonstring).take(15)
                    .joinToString("\n") { "${it.players.joinToString(" vs ")}: https://replay.pokemonshowdown.com/${it.id}" }
                    .ifEmpty { "Es wurde kein Kampf ${u2?.let { "zwischen $u1 und $it" } ?: "von $u1"} hochgeladen!" })
        } catch (ex: Exception) {
            iData.reply("Es konnte keine Verbindung zum Showdown-Server hergestellt werden!")
        }
    }

    private suspend fun getBody(u1: String, u2: String?) = httpClient.get(
        "https://replay.pokemonshowdown.com/search.json?user=${u1.toUsername()}".notNullAppend(u2?.let {
            "&user2=${it.toUsername()}"
        }).also { logger.info(it) }
    ).bodyAsText()


    private val logger = KotlinLogging.logger {}


    @Serializable
    data class Replay(val id: String, val players: List<String>, val uploadtime: Long, val format: String)
}
